app/helpers/pipeline_runs_helper.rb

```diff
@@ -221,4 +224,27 @@ def check_for_user_error(failed_stage)
     end
     [nil, nil]
   end
+
+  # Return all pipeline runs that have succeeded for given samples
+  # Only check the first pipeline run.
+  # samples should be an ActiveRecord relation
+  # If strict mode is turned on, error out even if one pipeline run did not succeed.
+  def get_succeeded_pipeline_runs_for_samples(samples, strict = false)
+    # Include the pipeline run for efficient queries.
+    samples = samples.includes(:pipeline_runs)
+
+    first_pipeline_runs = samples.map(&:first_pipeline_run)
```

Comment:
This is very slow... one query per sample.

Reply:
Hmm...I was hoping the `includes` would take care of this, but it looks like it doesn't. Any ideas how to make this faster?

Reply:
I was looking... i know it MySQL: `SELECT pr.id, pr.sample_id FROM (SELECT sample_id, MAX(created_at) AS max_created_at FROM pipeline_runs pr GROUP BY sample_id) AS x INNER JOIN pipeline_runs AS pr ON pr.sample_id=x.sample_id AND pr.created_at=x.max_created_at;`
Still looking for the best way to do it in rails...

Reply:
The only thing I can think of is to manually sort the pipeline runs in Ruby and then pick the correct one. So rewriting the `order_by` logic in Ruby. I'm assuming that if you call order_by, it will make a Rails query no matter what?